hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/security/client/ClientToAMTokenIdentifier.java

```diff
@@ -18,84 +18,191 @@
+            if (id != null) {
+                builder.setAppAttemptId(((ApplicationAttemptIdPBImpl) id).getProto());
+            }
+            if (client != null) {
+                builder.setClientName(client);
+            }
+            proto = builder.build();
+        }
+
+        public ApplicationAttemptId getApplicationAttemptID() {
+            if (!proto.hasAppAttemptId()) {
+                return null;
+            }
+            return new ApplicationAttemptIdPBImpl(proto.getAppAttemptId());
+        }
+
+        public String getClientName() {
+            return proto.getClientName();
+        }
+
+        public YarnSecurityTokenProtos.ClientToAMTokenIdentifierProto getProto() {
+            return proto;
+        }
+
+        @Override
+        public Text getKind() {
+            return KIND_NAME;
+        }
+
+        @Override
+        public UserGroupInformation getUser() {
+            String clientName = getClientName();
+            if (clientName == null) {
+                return null;
+            }
+            return UserGroupInformation.createRemoteUser(clientName);
+        }
+
+        @Override
+        public void write(DataOutput out) throws IOException {
+            out.write(proto.toByteArray());
+        }
+
+        @Override
+        public void readFields(DataInput in) throws IOException {
+            proto = YarnSecurityTokenProtos.ClientToAMTokenIdentifierProto.parseFrom((DataInputStream)in);
+        }
+    }
+
+    public class ByteArray extends ClientToAMTokenIdentifier {
+        private ApplicationAttemptId applicationAttemptId;
+        private Text clientName = new Text();
+
+        public ByteArray() {
+        }
+
+        public ByteArray(ApplicationAttemptId id, String client) {
+            this.applicationAttemptId = id;
+            this.clientName = new Text(client);
+        }
+
+        @Override
+        public void write(DataOutput out) throws IOException {
+            out.writeLong(this.applicationAttemptId.getApplicationId()
+                .getClusterTimestamp());
+            out.writeInt(this.applicationAttemptId.getApplicationId().getId());
+            out.writeInt(this.applicationAttemptId.getAttemptId());
+            this.clientName.write(out);
+        }
+
+        @Override
+        public void readFields(DataInput in) throws IOException {
+            this.applicationAttemptId =
+                ApplicationAttemptId.newInstance(
+                    ApplicationId.newInstance(in.readLong(), in.readInt()), in.readInt());
+            this.clientName.readFields(in);
+        }
+
+        public ApplicationAttemptId getApplicationAttemptID() {
+            return this.applicationAttemptId;
+        }
+
+        public String getClientName() {
+            return this.clientName.toString();
+        }
+
+        @Override
+        public Text getKind() {
+            return KIND_NAME;
+        }
+
+        @Override
+        public UserGroupInformation getUser() {
```

Comment:
no need to override this method, the one from the parent class relies on getClientName(). overriding getClientName() in subclasses is enough